proposals/nnnn-span-access-shared-contiguous-storage.md

```diff
@@ -0,0 +1,821 @@
+}
+```
+
+Note that we use a `_read` accessor for the subscript, a requirement in order to `yield` a borrowed non-copyable `Element` (see ["Coroutines"](#Coroutines).) This will be updated to a final syntax at a later time.
+
+##### Unchecked access to elements:
+
+The `subscript` mentioned above has always-on bounds checking of its parameter, in order to prevent out-of-bounds accesses. We also want to provide unchecked variants as an alternative for cases where bounds-checking is proving costly, such as in tight loops:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  // Unchecked subscripting and extraction
+
+  /// Accesses the element at the specified `position`.
+  ///
+  /// This subscript does not validate `position`; this is an unsafe operation.
+  ///
+  /// - Parameter position: The offset of the element to access. `position`
+  ///     must be greater or equal to zero, and less than `count`.
+  public subscript(unchecked position: Int) -> Element { _read }
+}
+```
+
+##### Index validation utilities:
+
+Every time `Span` uses a position parameter, it checks for its validity, unless the parameter is marked with the word "unchecked". The validation is performed with these functions:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  /// Return true if `index` is a valid offset into this `Span`
+  ///
+  /// - Parameters:
+  ///   - index: an index to validate
+  /// - Returns: true if `index` is a valid index
+  public func boundsContain(_ index: Int) -> Bool
+
+  /// Traps if `index` is not a valid offset into this `Span`
+  ///
+  /// - Parameters:
+  ///   - index: an index to validate
+  public func boundsPrecondition(_ index: Int)
+
+  /// Return true if `indices` is a valid range of offsets into this `Span`
+  ///
+  /// - Parameters:
+  ///   - indices: a range of indices to validate
+  /// - Returns: true if `indices` is a valid range of indices
+  public func boundsContain(_ indices: Range<Int>) -> Bool
+
+  /// Traps if `indices` is not a valid range of offsets into this `Span`
+  ///
+  /// - Parameters:
+  ///   - indices: a range of indices to validate
+  public func boundsPrecondition(_ indices: Range<Int>)
+}
+```
+
+Note: these function names are not ideal.
+
+##### Identifying whether a `Span` is a subrange of another:
+
+When working with multiple `Span` instances, it is often desirable to know whether one is a subrange of another. We include a function to determine whether this is the case, as well as a function to obtain the valid offsets of the subrange within the larger span:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  /// Returns true if the memory represented by `span` is a subrange of
+  /// the memory represented by `self`
+  ///
+  /// Parameters:
+  /// - span: a span of the same type as `self`
+  /// Returns: whether `span` is a subrange of `self`
+  public func contains(_ span: borrowing Self) -> Bool
```

Comment:
[Notes from in person discussions]

1. `span.isWithin(other)` appears to be a workable spelling for this. 

2. We also need to have an `isIdentical(to:)` method, to check equivalence. (`a.isIdentical(to: b) ==  a.isWithin(b) && b.isWithin(a)`, but we wouldn't want to spell/implement it that way.)

3. If the elements are bitwise copyable, then this operation needs to check alignment, which is a surprising complication, explaining why we'd want to ship this functionality in the stdlib. (This also needs to be called out within the docs.) 

    This may also turn into a solid argument against using the `isWithin` name, especially when considering the equivalent APIs on `Range` in point 5 below. (If so, perhaps something like `isSubspan` is the way to go.)

4. We should add equivalent members to the `UnsafeBufferPointer` types, to keep the stdlib coherent with itself, and to resolve some long standing omissions of basic functionality. 

   ```swift
   extension Unsafe[Mutable]BufferPointer {
     func isIdentical(to other: UnsafeBufferPointer<Element>) -> Bool
     func isIdentical(to other: UnsafeMutableBufferPointer<Element>) -> Bool
     func isWithin(_ other: UnsafeBufferPointer<Element>) -> Bool
     func isWithin(_ other: UnsafeMutableBufferPointer<Element>) -> Bool
   }
   ```

5. `isWithin` ought to also be added to the range types:

   ```swift
   extension Range {
     func isWithin(other: Range) -> Bool
     func isWithin(other: ClosedRange) -> Bool
   }
   extension ClosedRange {
     func isWithin(other: Range) -> Bool
     func isWithin(other: ClosedRange) -> Bool
   }
   ```

   (These types already provide the functionality of `isIdentical` in their `==` operation.)

   The notion of unaligned bounds makes this somewhat dangerous -- `Range<UnsafePointer<T>>.isWithin(_:)` would sometimes return different results than an `isWithin` call on `UnsafeBufferPointer` or `Span` values over the same region of memory.

6. The preexisting `[Closed]Range.overlaps` hints at a need to provide `[Raw]Span.overlaps` and `Unsafe[Mutable][Raw]BufferPointer.overlaps` operations. (As well as `intersect` and `union`/`merge`/`combine` operations.) This does not seem to be an urgent need though.